subsys/net/lib/conn_mgr/conn_mgr_connectivity.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for managing L2 layers that have separated post-admin-up connect
+ *	  or associate phases
+ */
+
+#include <zephyr/net/net_if.h>
+
+/**
+ * @brief Retrieves the conn_mgr binding struct for a provided iface if it exists.
+ *
+ * Bindings for connectivity implementations with missing API structs are ignored.
+ *
+ * @param iface - bound network interface to obtain the binding struct for.
+ * @return struct conn_mgr_binding* Pointer to the retrieved binding struct if it exists,
+ *	   NULL otherwise.
+ */
+static inline struct conn_mgr_binding *conn_mgr_if_get_binding(struct net_if *iface)
+{
+	STRUCT_SECTION_FOREACH(conn_mgr_bound, binding) {
+		if (iface == conn->iface) {
+			if (conn->impl->api) {
+				return binding;
+			}
+			return NULL;
+		}
+	}
+	return NULL;
+}
+
+int conn_mgr_if_connect(struct net_if *iface)
+{
+	NET_DBG("iface %p connect", iface);
+
+	const struct conn_mgr_binding *binding = conn_mgr_if_get_binding(iface);
+
+	if (!binding) {
+		return -ENOTSUP;
+	}
+
+	const struct conn_mgr_conn_api *api = binding->impl->api;
+
+	if (!api->connect) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	int status = 0;
+
+	if (!net_if_flag_is_set(iface, NET_IF_UP)) {
+		status = -ESHUTDOWN;
+		goto out;
+	}
+
+	status = api->connect(binding);
+
+out:
+	k_mutex_unlock(&lock);
+
+	return status;
+}
+
+int conn_mgr_if_disconnect(struct net_if *iface)
+{
+	NET_DBG("iface %p disconnect", iface);
+
+	const struct conn_mgr_binding *binding = conn_mgr_if_get_binding(iface);
+
+	if (!binding) {
+		return -ENOTSUP;
+	}
+
+	const struct conn_mgr_conn_api *api = binding->impl->api;
+
+	if (!api->disconnect) {
+		return -ENOTSUP;
+	}
+
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	int status = 0;
+
+	if (!net_if_flag_is_set(iface, NET_IF_UP)) {
+		status = -EALREADY;
+		goto out;
+	}
+
+	status = api->disconnect(binding);
+
+out:
+	k_mutex_unlock(&lock);
+
+	return status;
+}
+
+bool conn_mgr_if_is_bound(struct net_if *iface)
+{
+	struct conn_mgr_binding *binding = conn_mgr_if_get_binding(iface);
+
+	return binding != NULL;
+}
+
+int conn_mgr_if_get_opt(struct net_if *iface, int optname, void *optval, size_t *optlen)
+{
+	if (!optlen) {
+		return -EINVAL;
+	}
+
+	if (!optval) {
+		*optlen = 0;
+		return -EINVAL;
+	}
+
+	const struct conn_mgr_binding *binding = conn_mgr_if_get_binding(iface);
+
+	if (!binding) {
+		*optlen = 0;
+		return -ENOTSUP;
+	}
+
+	const struct conn_mgr_conn_api *api = binding->impl->api;
+
+	if (!api->get_opt) {
+		*optlen = 0;
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
```

Comment:
This mutex doesn't exist

Reply:
Perhaps instead of a global mutex, we should define a mutex as a part of the binding structure? There should be no interference between operations on separate bindings. And if there is, it should be handled at the connectivity implementation level.

Reply:
I was thinking the same thing, a per-binding mutex would make more sense